portfolio/src/main/client/src/components/controls/labeled-input.ts

```diff
@@ -0,0 +1,85 @@
+import {htmlElement, htmlFragment} from '@src/util/html';
+
+/**
+ * Options for labeled inputs.
+ */
+export interface LabeledInputOptions {
+  /**
+   * The label. Will be shown as a placeholder when the textbox is empty, and
+   * shown as a label otherwise.
+   */
+  label: string;
+
+  /**
+   * The ID to apply to the input or textarea.
+   */
+  id: string;
+
+  /**
+   * Whether the label will disappear automatically when the input is empty.
+   * Does not apply to textareas, since they do not support placeholders.
+   * Defaults to true.
+   */
+  soft?: boolean;
+
+  /**
+   * The name of the input or textarea.
+   */
+  name?: string;
+
+  /**
+   * The class name of the input or textarea.
+   */
+  className?: string;
+
+  /**
+   * The type of input to create.
+   */
+  type?: 'text' | 'number' | 'email' | 'textarea';
+
+  /**
+   * A predefined value to put in the input or textarea.
+   */
+  value?: string;
+}
+
+/**
+ * @returns A labeled input, which is just a label followed by an input or
+ * textarea. This is designed to be used with textual inputs: when the input is
+ * empty, the label may disappear and be used as a placeholder instead.
+ */
+export const LabeledInput =
+  ({
+    name, id, className, label, type, value, soft,
+  }: LabeledInputOptions): Node[] => {
+    const labelEl: HTMLLabelElement =
+      htmlElement`<label for="${id}" class="input-label">${label}</label>`;
+
+    if (soft !== false) labelEl.classList.add('soft');
```

Comment:
Doing it the way you suggested has the same effect as what I was doing: unless you specify the `soft` option explicitly, it will default to `true`. If I don't do anything, `soft` will default to `undefined`, and I could treat that as `false`, but then I would have to specify `soft` every time I want to use `LabeledInput` which gets on my nerves because they will probably always be soft.

Reply:
I changed it, anyway.

Reply:
I understand the frustration. The difference is defining this at the interface level (e.g. anyone using this interface should implement it such that `soft === undefined` is the same as `soft === true`) vs at the method level (where it's just this method that overrides the default value of soft).